File: linked-list/24SwapNodesInPairs.java
public class SwapNodesInPairs {
    class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public ListNode swapPairs(ListNode head) {
        if(head == null || head.next == null) return head;
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode prev = dummy;
        while(head != null && head.next != null) {
            ListNode first = head, second = head.next, next = second.next;
            prev.next = second;
            second.next = first;
            first.next = next;
            
            prev = first;
            head = first.next;
        }
        return dummy.next;
    }

    public ListNode recursive(ListNode head) {
        if(head != null || head.next != null) return head;
        ListNode first = head, second = head.next;
        first.next = recursive(second.next);
        second.next = first;
        return second;
    }
}

File: leetcode75/l3/medium/24SwapNodesInPairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Time/Space: O(N)
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(-1)
        dummy.next = head
        
        prev = dummy
        while head and head.next:
            first = head
            second = head.next
            
            prev.next = second
            first.next = second.next
            second.next = first
            
            prev = first
            head = first.next
        return dummy.next

        # # If the list has no node or has only one node left.
        # if not head or not head.next:
        #     return head

        # # Nodes to be swapped
        # first_node = head
        # second_node = head.next

        # # Swapping
        # first_node.next  = self.swapPairs(second_node.next)
        # second_node.next = first_node

        # # Now the head is the second node
        # return second_node